Create trusted-proof-serving.md for proof generation (#81)
midnightntwrk/midnight-improvement-proposals: mps/trusted-proof-serving.md

---

> MPS: ???

Reviewer: ```suggestion
MPS: 0004
```

### UC3: Network-Wide Proof Service with Verifiable Integrity

**Scenario**: An ecosystem/community member operates a proof server for Midnight community  users. Community members and auditors want to verify that the Foundation's server is running authentic, unmodified code.

**Current Approach**: The proof server operator publishes the server's source code and claims it runs unmodified. Users must trust this claim.

**Limitations**: Source code publication proves nothing about what binary is actually running in production. A modified binary could be deployed that exfiltrates data while remaining source-compatible.

**Desired Outcome**: The server produces a cryptographic attestation document on demand. Any client can verify: (1) the server runs inside a genuine TEE, (2) the binary matches a published, reproducible build (identified by PCR0/hash), and (3) the attestation is fresh (not replayed). Verification requires only the attestation document, the TEE vendor's root certificate, and the published build measurements — no trust in the server operator.

### UC4: Multi-Cloud Operator Deployment

> **Scenario**: An enterprise operating Midnight infrastructure uses a multi-cloud strategy (AWS primary, GCP failover). They need proof generation in multip;e environments with equivalent security guarantees.

Reviewer: ```suggestion
**Scenario**: An enterprise operating Midnight infrastructure uses a multi-cloud strategy (AWS primary, GCP failover). They need proof generation in multiple environments with equivalent security guarantees.
```

5. **Reproducible builds for PCR verification**: The build process must be deterministic so that anyone can rebuild the binary, compute the expected PCR0 (code measurement), and compare it against the published value. This closes the gap between "open source" and "verifiably running this open source."

### Requirements

- **Performance**: Proof generation throughput ≥ 80 proofs/sec at 16 workers on 32-core hardware. p50 latency < 250ms, p99 < 600ms for standard ZSwap proofs. Cold start to ready (including parameter pre-fetch) < 60 seconds.
- **Security**: API keys hashed (never stored plaintext). TLS by default for all communications. No persistent storage inside the TEE. Per-IP rate limiting to prevent abuse. Graceful authentication with zero-downtime key rotation.
- **Privacy**: Witness data and proving keys exist only in TEE-encrypted memory during proof generation. No logging of witness content. Memory is released after each proof. No persistent storage that could be forensically recovered.

> - **Compatibility**: Proof output must be identical to client-side generation — the on-chain verifier cannot distinguish server-generated from client-generated proofs. Binary serialization format must match the midnight SDK's tagged serialization.

Reviewer: ```suggestion
- **Compatibility**: Proof output must be identical to client-side generation — the on-chain verifier cannot distinguish server-generated from client-generated proofs. Binary serialization format must match the Midnight SDK's tagged serialization.
```
